lib/irq.c

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright (c) 2019, Xilinx Inc. and Contributors. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include <errno.h>
+#include <metal/irq.h>
+#include <metal/list.h>
+#include <metal/utilities.h>
+#include <metal/atomic.h>
+
+/** Soft IRQs mapping it stores which IRQ triggers the soft IRQ */
+struct metal_irq_map {
+	int source_irq_base; /**< Source IRQ which triggers the soft IRQ */
+	int target_irq_base; /**< target soft IRQ */
+	int irq_num; /**< number of IRQs of the mapping */
+};
+
+/** Indicate which IRQ has been allocated */
+#define METAL_IRQ_MAX 2048
+static unsigned long
+metal_irq_allocated[metal_div_round_up(METAL_IRQ_MAX, METAL_BITS_PER_ULONG)];
+
+/** List of registered IRQ controller */
+static METAL_DECLARE_LIST(irq_cntrs);
+
+#define METAL_IRQ_MAX_MAPPINGS 1
+static struct metal_irq_map irq_maps[METAL_IRQ_MAX_MAPPINGS];
+
+int metal_irq_allocate(int irq_base, int irq_num)
+{
+	static atomic_flag flag;
+	int ret, i;
+
+	while(atomic_flag_test_and_set(&flag) != 0);
+	if (irq_base != METAL_IRQ_ANY) {
+		if (metal_bitmap_is_bit_set(metal_irq_allocated, irq_base)
+		    != 0) {
+			ret = -EINVAL;
+			goto err;
+		}
+		if ((irq_base + irq_num) >= METAL_IRQ_MAX) {
+			ret = -EINVAL;
+			goto err;
+		}
+	} else {
+		irq_base = metal_bitmap_next_clear_bit(metal_irq_allocated,
+						       0, METAL_IRQ_MAX);
+		if (irq_base == METAL_IRQ_MAX) {
+			ret = -EINVAL;
+			goto err;
+		}
+	}
+
+	for (i = irq_base; i < irq_base + irq_num; i++) {
+		metal_bitmap_set_bit(metal_irq_allocated, i);
+	}
+	ret = irq_base;
+
+err:
+	atomic_flag_clear(&flag);
+	return ret;
+}
+
+void metal_irq_register_controller(struct metal_irq_controller *cntr)
+{
+	struct metal_list *node;
+	static atomic_flag flag;
+
+	if (cntr == NULL)
+		return;
+	while(atomic_flag_test_and_set(&flag) != 0);
+	metal_list_for_each(&irq_cntrs, node) {
+		if (node == &cntr->node) {
+			atomic_flag_clear(&flag);
+			return;
+		}
+	}
+	metal_list_add_tail(&irq_cntrs, &cntr->node);
+	atomic_flag_clear(&flag);
+}
+
+int metal_irq_get_softirq(int irq)
+{
+	int i;
+
+	for(i = 0; i < METAL_IRQ_MAX_MAPPINGS; i++) {
+		int offset;
+		int source_irq_base;
+
+		source_irq_base = irq_maps[i].source_irq_base;
+		offset = irq - source_irq_base;
+
+		if (irq >= source_irq_base && offset < irq_maps[i].irq_num) {
+			return irq_maps[i].target_irq_base + offset;
+		}
+	}
+	return -EINVAL;
+}
+
+struct metal_irq_controller *metal_irq_get_controller(int irq)
```

Comment:
This function does not lock the list from changing while it is iterating it.  This could cause the list to be corrupted if we allow multiple threads to dynamically add interrupt controllers.  I cannot think of any use case where we need to add interrupt controllers dynamically or by multiple threads.  Is there a use case for this?  If there is no such use case I suggest to remove the locking from metal_irq_register_controller().

If there are many interrupt controllers then using a list can be inefficient, so we might want to change the list to an array at some point...

Reply:
Will remove metal_irq_get_controller() function from user API. As it is only used internally to call interrupt controller to enable/disable interrupt. It was also used to set the softirq to pending, but this will be covered by a new softirq function. About the locking, agree that we can limit to say that the controller needs to be registered at the initialization time. And not supposed to register controllers by multiple threads. And then remove the locking.